### PR TITLE
Persist finalized state and be able to load it from DB

### LINKED ISSUE #136: Persist finalized state and be able to load it from DB
Fruzhin should persist the state trie and new finalized state to the DB everytime it verifies a new block.(Think about whether it should overwrite the old state or create a new one)

Also think about whether we should save peer info and be able to load it from the DB, as well as other Host information

### PR DESCRIPTION
**Pull request by ablax**
* Save and load synced state on commit
* Save lightSyncState and load from db
* Save the latest trie in db

Closes #136

### CODE REVIEW

File: src/main/java/com/limechain/network/Network.java [hunk 1]
```diff
@@ -85,7 +85,6 @@ public Network(ChainService chainService, HostConfig hostConfig, KVRepository<St
         this.initializeProtocols(chainService, hostConfig, repository, cliArgs);
         this.bootNodes = chainService.getGenesis().getBootNodes();
         this.chain = hostConfig.getChain();
-        SyncedState.getInstance().setNetwork(network);
```
Comment (vikinatora) [inline src/main/java/com/limechain/network/Network.java:88]: Why is this removed?

File: src/main/java/com/limechain/network/protocol/blockannounce/BlockAnnounceEngine.java [hunk 2]
```diff
@@ -68,6 +68,7 @@ private void handleBlockAnnounce(byte[] msg, PeerId peerId) {
                 " with hash:0x" + HexUtils.toHexString(announce.getHeader().getHash()) +
                 " parentHash:" + announce.getHeader().getParentHash() +
                 " stateRoot:" + announce.getHeader().getStateRoot());
+        //TODO: Should update Trie (and merkle proofs?) and save them in db
```
Comment (vikinatora) [inline src/main/java/com/limechain/network/protocol/blockannounce/BlockAnnounceEngine.java:71]: _**I think that**_ if the state root of the header has changed the host should send out a state request for the block hash(since light client can't process extrinsics) and update it's trie.

IMO this should be in a separate task in order to not bloat this PR

File: src/main/java/com/limechain/sync/warpsync/SyncedState.java [hunk 3]
```diff
@@ -120,5 +106,60 @@ private synchronized void updateState(CommitMessage commitMessage) {
         lastFinalizedBlockHash = commitMessage.getVote().getBlockHash();
         lastFinalizedBlockNumber = commitMessage.getVote().getBlockNumber();
         log.log(Level.INFO, "Reached block #" + lastFinalizedBlockNumber);
+        saveState();
+    }
+
+    public void saveState() {
```
Comment (vikinatora) [inline src/main/java/com/limechain/sync/warpsync/SyncedState.java]: let's change the function name to `persistState`

File: src/main/java/com/limechain/storage/DBConstants.java [hunk 4]
```diff
@@ -9,4 +9,19 @@ public class DBConstants {
...
+    /**
+     * Key under which the --latest-- state trie proof is stored
+     * TODO: Currently only the latest loaded in sync is stored
+     */
+    public static final String STATE_TRIE_PROOF = "stateTrieProof";
+    /**
+     * Key under which the --latest-- state trie root state is stored
+     * TODO: Currently only the latest loaded in sync is stored
+     */
+    public static final String STATE_TRIE_ROOT_STATE = "stateTrieRootState";
```
Comment (vikinatora) [inline src/main/java/com/limechain/storage/DBConstants.java]: let's rename it to `STATE_TRIE_ROOT_HASH`

File: src/main/java/com/limechain/storage/DBConstants.java [hunk 5]
```diff
@@ -9,4 +9,19 @@ public class DBConstants {
...
     public static final String GENESIS_KEY = "genesis";
+    /**
+     * Key under which the latest sync state is stored
+     */
+    public static final String SYNC_STATE_KEY = "syncState";
+    /**
+     * Key under which the --latest-- state trie proof is stored
+     * TODO: Currently only the latest loaded in sync is stored
+     */
+    public static final String STATE_TRIE_PROOF = "stateTrieProof";
```
Comment (vikinatora) [inline src/main/java/com/limechain/storage/DBConstants.java]: let's change it to `STATE_TRIE_MERKLE_PROOF`

**Review by vikinatora — CHANGES_REQUESTED** [review]
Some minor changes are requested

File: src/main/java/com/limechain/network/Network.java (hunk 1 above)
Comment (ablax) [reply src/main/java/com/limechain/network/Network.java:88]: It's not used anywhere

File: src/main/java/com/limechain/network/protocol/blockannounce/BlockAnnounceEngine.java (hunk 2 above)
Comment (ablax) [reply src/main/java/com/limechain/network/protocol/blockannounce/BlockAnnounceEngine.java:71]: I agree that it should be in separate task, that's why I left it with todo

File: src/main/java/com/limechain/sync/warpsync/WarpSyncMachine.java [hunk 6]
```diff
@@ -64,7 +64,7 @@ public void start() {
         final Hash256 initStateHash;
         if (this.chainService.getGenesis().getLightSyncState() != null) {
             this.syncedState.setRepository(repository);
-            boolean stateLoaded = this.syncedState.loadState();
+            boolean stateLoaded = !chainService.isLocalChain() && this.syncedState.loadState();
```
Comment (vikinatora) [inline src/main/java/com/limechain/sync/warpsync/WarpSyncMachine.java]: Not sure if this is expected behaviour. In most cases this would be the intuitive solution but it would confuse some users that want to restart their Host and continue with the sync.

I'd remove `!chainService.isLocalChain()`. If users want to reset their DB, they can use the `-dbc` flag

File: src/main/java/com/limechain/network/Network.java (hunk 1 above)
Comment (vikinatora) [reply src/main/java/com/limechain/network/Network.java:88]: I answered my own question - it's not used. You can resolve this comment

**Review by vikinatora — APPROVED** [review]
